### PR TITLE
MCR-2987 Add resolver to call arbitrary static Java functions in XSLT3 stylesheets

### PR DESCRIPTION
**Pull request by Possommi**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2987).

### CODE REVIEW

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java [hunk 1]
```diff
@@ -246,6 +248,7 @@ private HashMap<String, URIResolver> getResolverMapping() {
         supportedSchemes.put("https", restResolver);
         supportedSchemes.put("file", new MCRFileResolver());
         supportedSchemes.put("cache", new MCRCachingResolver());
+        supportedSchemes.put("function", new MCRFunctionResolver());
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: I would like to have a more precise name than simply `function:`
What about `java:` or `java-call:`

Comment (Possommi) [reply mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: In xed forms one can use `call-java`. Maybe we should stick with that.

Comment (rsteph-de) [reply mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: sounds good for me

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java [hunk 2]
```diff
@@ -1826,4 +1829,49 @@ public Source resolve(String href, String base) throws TransformerException {
...
+    private static class MCRFunctionResolver implements URIResolver {
+        @Override
+        public Source resolve(String href, String base) throws TransformerException {
+            LOGGER.debug("Resolving {}", href);
+
+            String[] parts = href.split(":");
+            String className = parts[1];
+            String methodName = parts[2];
+
+            int numParams = parts.length - 3;
```
Comment (sebhofmann) [inline mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: Could all be simplified by:

`Arrays.stream(parts).skip(3).toArray(String[]::new)`

Comment (Possommi) [reply mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: @sebhofmann using now the fancy stuff

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java (hunk 1 above)
Comment (Possommi) [reply mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: @rsteph-de updated schema to `call-java`

Comment (rsteph-de) [reply mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: Cool!

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java [hunk 3]
```diff
@@ -246,6 +248,7 @@ private HashMap<String, URIResolver> getResolverMapping() {
         supportedSchemes.put("https", restResolver);
         supportedSchemes.put("file", new MCRFileResolver());
         supportedSchemes.put("cache", new MCRCachingResolver());
+        supportedSchemes.put("call-java", new MCRFunctionResolver());
```
Comment (toKrause) [inline mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: Should be `callJava`

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java [hunk 4]
```diff
@@ -1826,4 +1829,44 @@ public Source resolve(String href, String base) throws TransformerException {
...
+     * <br/><br/>
+     * <strong>Invocation</strong>
+     * <pre><code>function:&lt;class name&gt;:&lt;method name&gt;:&lt;param1&gt;:&lt;param2&gt;</code></pre>
+     * <br/>
+     * <strong>Example</strong>
+     * <pre><code>function:de.uni_jena.thunibib.user.ThUniBibUtils:getLeadId:id_connection:foobar;</code></pre>
+     *
+     * @author shermann (Silvio Hermann)
+     * */
+    private static class MCRFunctionResolver implements URIResolver {
```
Comment (toKrause) [inline mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: Should be a separate class file and included with property `MCR.URIResolver.ModuleResolver.callJava=org.mycore.`...`.MCRFunctionResolver`

**Review by toKrause — CHANGES_REQUESTED** [review]

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java (hunk 3 above)
Comment (toKrause) [inline mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: Needs to be removed, now that the resolver is included by the property `MCR.URIResolver.ModuleResolver.callJava`

**Review by toKrause — CHANGES_REQUESTED** [review]

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java (hunk 3 above)
Comment (Possommi) [reply mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: mea culpa

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java [hunk 5]
```diff
@@ -34,6 +34,7 @@
 import java.time.Instant;
```
Comment (toKrause) [inline mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: One more thing, ... all the changes in this class now seem to be formatting/indentation changes and an `import` that now seem unnecessary. I think, all those changes can be dropped and this class be removed from the PR.

**Review by toKrause — CHANGES_REQUESTED** [review]
One more thing, sorry, ...

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java (hunk 5 above)
Comment (rsteph-de) [reply mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: Good point! We have already formatted the main branch to prepare the release, so changes in 2022.06 should be as small as possible.

Comment (Possommi) [reply mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: done

**Review by toKrause — APPROVED** [review]

**Comment by Possommi** [comment]
Currently in one of our applications arbitrary code can be invoked via url (sourceURI in xeditor)

http://localhost:9080/collections/editor-start.xed?type=cbu&action=create&editor.layout=20.0&sourceUri=callJava:org.mycore.common.xml.MCRXMLFunctions:trim:foo&cancelUrl=http://localhost:9080/collections/content/main/search/searchPica.xml

This resolver should not be active by default.

**Pull request merged**